Returns correct result to Mod(3*i, 2).

m-agboola opened the pull request:
modified the mod.py to return correct answer to Mod(3*i, 2).
added a test (All as suggested by @smichr )

Fixes #15493 

Earlier
` sympify(3*k%2)
Mod(3*k,2)`

Now
` sympify(3*k%2)
Mod(k,2)`

 **Release Notes**
<!-- BEGIN RELEASE NOTES -->
* functions
  * fixed a bug in mod 
  * added a test
<!-- END RELEASE NOTES -->

m-agboola committed:
correct Mod(3*i, 2) result, Fix #15493

smichr commented on sympy/core/mod.py:
```diff
@@ -117,15 +117,17 @@ def doit(p, q):
                 return cls(net, q)
 
-        elif isinstance(p, Mul):
+        if isinstance(p, Mul):
```
import `denom` from radsimp and then make this `if isinstance(p, Mul) and denom(p) == 1:` since only if a and b are coprime will `Mod(a/b,n)==Mod(a,n)*Mod(1/b,n)`.

smichr commented on sympy/core/mod.py:
```diff
@@ -117,15 +117,17 @@ def doit(p, q):
                 both_l[isinstance(arg, cls)].append(arg)
 
-            if mod_l and all(inner.args[1] == q for inner in mod_l):
+            was = non_mod_l[:]
```
these 3 lines can be
```
non_mod_l = [cls(x, q) for x in n]
changed = not all(isinstance(i, cls) for i in non_mod_l)
```
And a test that `assert Mod(e/2, 2) != 0` where `e = symbols('e', even=True)` should be added.

m-agboola replied:
Hi @smichr 
I just updated the pull request after including the changes suggested above.

m-agboola committed:
Returns correct result to Mod(3*i, 2) and improve the result of Mod(a/b,n). Fix issue #15493.

m-agboola committed:
Fixed issue_#15493

m-agboola committed:
Fixes issue_#1503

avishrivastava11 commented:
@m-agboola  
The changes you have made will indicate that `Mod(var('e',even=True)/2,2)==0` but that should remain unevaluated. Make sure you make that work correctly before going ahead further.

m-agboola committed:
Revert Mod(var('k',even=True)/2,2) == 0

smichr committed:
don't import denom at top

Sometimes you can't put an import at the type of the file because of the order that imports are done.

smichr commented:
Please add the test, 
```
e = symbols('e', even=True)
assert Mod(e/2, 2).subs(e, 6) == Mod(3, 2)
```

smichr committed:
relocate denom to correct function

m-agboola committed:
Import radsimp correctly and add a test

m-agboola committed:
Merge branch 'issue_#15493' of https://github.com/mukhtarAgboola/sympy into issue_#15493

m-agboola commented:
Hi, I just implemented the suggested changes stated above.

smichr commented:
After adding the test I suggested, now take a look at the results of split 3 and 4 (the Travis results) and 

1. see what the modified code is giving, e.g. what does `(x - 3.3) % 1` given now; it was formerly `Mod(1.*x + 1-.3, 1)`
1. see if it makes sense or if it represents another corner case that should be avoided by the code and 
1. make the appropriate update to test or code

m-agboola commented:
`(x - 3.3) % 1` still gives `Mod(1.*x + .7, 1)` which is equivalent to `Mod(1.*x + 1-.3, 1)`. I'm not sure I understand why Travis is failing.

smichr commented:
> still gives `Mod(1.*x + .7, 1)`

Make sure you are in your branch and not master; if you are using Windows and switched branch then you have to restart the interactive session (if that is how you are checking this).

m-agboola commented:
Yes, they are not the same, `Mod(1.*x + .7, 1)` give `0.7000000` which probably assumes `x` is a whole number. It happens only when the coefficient of `x` is a whole number but I still don't know how to fix this
